core/drivers/clk/sam/at91_pmc.c

```diff
@@ -114,3 +120,167 @@ struct pmc_data *pmc_data_allocate(unsigned int ncore, unsigned int nsystem,
 
 	return pmc_data;
 }
+
+#ifdef CFG_PM_ARM32
+static uint8_t registered_ids[PMC_MAX_IDS];
+static uint8_t registered_pcks[PMC_MAX_PCKS];
+
+static struct
+{
+	uint32_t scsr;
+	uint32_t pcsr0;
+	uint32_t uckr;
+	uint32_t mor;
+	uint32_t mcfr;
+	uint32_t pllar;
+	uint32_t mckr;
+	uint32_t usb;
+	uint32_t imr;
+	uint32_t pcsr1;
+	uint32_t pcr[PMC_MAX_IDS];
+	uint32_t audio_pll0;
+	uint32_t audio_pll1;
+	uint32_t pckr[PMC_MAX_PCKS];
+} pmc_cache;
+
+/*
+ * As Peripheral ID 0 is invalid on AT91 chips, the identifier is stored
+ * without alteration in the table, and 0 is for unused clocks.
+ */
+void pmc_register_id(uint8_t id)
+{
+	int i = 0;
+
+	for (i = 0; i < PMC_MAX_IDS; i++) {
+		if (registered_ids[i] == 0) {
+			registered_ids[i] = id;
+			break;
+		}
+		if (registered_ids[i] == id)
+			break;
+	}
+}
+
+/*
+ * As Programmable Clock 0 is valid on AT91 chips, there is an offset
+ * of 1 between the stored value and the real clock ID.
+ */
+void pmc_register_pck(uint8_t pck)
+{
+	int i = 0;
+
+	for (i = 0; i < PMC_MAX_PCKS; i++) {
+		if (registered_pcks[i] == 0) {
+			registered_pcks[i] = pck + 1;
+			break;
+		}
+		if (registered_pcks[i] == pck + 1)
+			break;
+	}
+}
+
+static void pmc_suspend(void)
+{
+	int i = 0;
+	uint8_t num = 0;
+	vaddr_t pmc_base = at91_pmc_get_base();
+
+	pmc_cache.scsr = io_read32(pmc_base + AT91_PMC_SCSR);
+	pmc_cache.pcsr0 = io_read32(pmc_base + AT91_PMC_PCSR);
+	pmc_cache.uckr = io_read32(pmc_base + AT91_CKGR_UCKR);
+	pmc_cache.mor = io_read32(pmc_base + AT91_CKGR_MOR);
+	pmc_cache.mcfr = io_read32(pmc_base + AT91_CKGR_MCFR);
+	pmc_cache.pllar = io_read32(pmc_base + AT91_CKGR_PLLAR);
+	pmc_cache.mckr = io_read32(pmc_base + AT91_PMC_MCKR);
+	pmc_cache.usb = io_read32(pmc_base + AT91_PMC_USB);
+	pmc_cache.imr = io_read32(pmc_base + AT91_PMC_IMR);
+	pmc_cache.pcsr1 = io_read32(pmc_base + AT91_PMC_PCSR1);
+
+	for (i = 0; registered_ids[i]; i++) {
+		io_write32(pmc_base + AT91_PMC_PCR,
+			   (registered_ids[i] & AT91_PMC_PCR_PID_MASK));
+		pmc_cache.pcr[registered_ids[i]] = io_read32(pmc_base +
+							     AT91_PMC_PCR);
+	}
+	for (i = 0; registered_pcks[i]; i++) {
+		num = registered_pcks[i] - 1;
+		pmc_cache.pckr[num] = io_read32(pmc_base + AT91_PMC_PCKR(num));
+	}
+}
+
+static bool pmc_ready(vaddr_t pmc_base, unsigned int mask)
+{
+	uint32_t status;
+
+	status = io_read32(pmc_base + AT91_PMC_SR);
+
+	return (status & mask) == mask;
+}
+
+static void pmc_resume(void)
+{
+	int i = 0;
+	uint8_t num = 0;
+	uint32_t tmp = 0;
+	vaddr_t pmc_base = at91_pmc_get_base();
+	uint32_t mask = AT91_PMC_MCKRDY | AT91_PMC_LOCKA;
+
+	tmp = io_read32(pmc_base + AT91_PMC_MCKR);
+	if (pmc_cache.mckr != tmp)
```

Comment:
I'll check that yes.

Reply:
After looking more carefully at the datasheet, the worst thing that could happen is to run at a different frequency that initially. I would guess a lower one since this is often what is done by the firmware. To be safer, I would propose to panic in such cases, better be safe than sorry.

Reply:
OK